Add a common test file shared among all containers

trepik opened the pull request:
This file contains common bits for container tests, and should be sourced by each specific container test. It is created to keep the code on one place and not spread through the repositories.

praiskup commented on tests.sh:
```diff
@@ -0,0 +1,113 @@
+  rmdir $CID_FILE_DIR
+  : "Done."
+}
+trap cleanup EXIT SIGINT
```
As this is now library, we should explicitly ask for the cleanup action;  so I would suggest you to have some `enable_auto_cleanup` method.

praiskup commented:
Thanks for working on it!

- my first thought is as-clean-API-as-possible... please design naming conventions for the api calls and use that;  e.g. `cictests_*` prefix (ci-container-tests) for methods

- document how to use it at the beginning of the file

- (optional) provide one pull request against some container that would depend on this feature, that would simplify testing of the proposal :)

trepik commented:
- added prefix to all function names
- added extra function enabling the cleanup
- documentation will follow once it's working with more containers
- one can check the functionality with Mongo 3.2 container from my [mongo-container](https://github.com/trepik/mongodb-container) fork

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+# run_doc_test <words_to_look_for_in_the_doc>"
+
+# Container CI tests
+# abbreviated as "concit"
```
Could be OK, but @pkubatrh, @omron93, @hhorak FYI.  Ideas?

I have no strong preference, but maybe `ct_` prefix could be enough  (Container Testing).

Once we claim we provide "library", it will be much harder to migrate to different function prefix.  So we should rather agree that we use pretty namespace.

pkubatrh replied:
+1 for `ct_`. The shorter the better.

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+
+function concit_cleanup() {
+  for cid_file in $CID_FILE_DIR/* ; do
+    CONTAINER=$(cat $cid_file)
```
The CONTAINER variable is not `local`, which creates shell variable pollution.

I would encourage you to document shell functions like:
```bash
# foo_action [ARGS...]
# --------------------
# Does "something" with all ARGS. Set's $X to $Z value.
# Defines: $X
# Uses: $Y
foo_actions ()
{
    X=$Y
    for arg ; do
        something "$arg"
    done
    local Z=baz # Z is not visible outside...
}
```

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+  # container will keep running so we kill it with SIGKILL to make sure
+  # timeout returns a non-zero value.
+  set +e
+  timeout -s SIGTERM --preserve-status 10s docker run --rm "$@" $IMAGE_NAME
```
I'm bit confused reading this call....   It returns '1' if the command timeouts, otherwise it returns 0.

So the name should be '*assert_container_start_timeouts' maybe?

hhorak replied:
I don't think the timeout is important here, it is just an implementation.. The result should be simply 0 if the container fails, 1 if it does not.

hhorak replied:
I'd also increase the timeout to 30s since if the container is expected to fail after the initialization of the database, we can timeout too soon.. as mentioned above, if the container works (fails) as expected, it will fail much sooner in most cases..

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+  # create container with a cidfile in a directory for cleanup
+  docker run ${CONTAINER_ARGS:-} --cidfile="$cid_file" -d "$@" $IMAGE_NAME
+  : "Created container $(cat $cid_file)"
+  concit_wait_for_cid
```
Could `wait_for_cid` method have an argument, rather than global `$cid_file` variable?

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+  concit_wait_for_cid
+}
+
+function concit_scl_usage() {
```
API of this function is really unclear..  Seems like it is `grep_container_output`, but I'm unsure TBH

hhorak replied:
`grep_container_output` makes sense to me.

hhorak replied:
Actually, I've realized that this function really tests whether `scl enable` works... that's why there are three calls done: `docker run`, `docker exec` with `bash` and `docker exec` with `sh`... But something like `grep_container_output` may be handy as well.

trepik replied:
If you wanted a function `grep_container_command_output` you'd probably run the docker only once and not 3 times so therefore keeping the name. Once there is a better implementation of testing scl_usage it could be replaced.

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+  fi
+}
+
+function concit_run_doc_test() {
```
This is actually a test, not library-call.  It should be obvious from the function name ... Proposals?

hhorak replied:
`ct_` prefix for library functions and `ctest_` or `run_ctest_` prefix for tests? Or just simply `run_test_` prefix for tests?

praiskup replied:
`ct_` for lib and `ctest_` for tests sounds OK to me.

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,142 @@
+  : "  Success!"
+}
+
+function concit_run_all_tests() {
```
Could we leave this unsolved by container library, and do it as separate PR with something more "revolutionary"?

My thought would be to have separate file like `run-tests` for this method...  e.g.  And perhaps split particular tests into separate files too in future (so we can specify list of tests to be performed via arguments, or make variables).

praiskup commented:
Thanks a lot for the work.  My thought is to not necessarily move everything now, but start with smaller set of "clearly useful and well defined API functions".  Even if we move only five methods, it is good first step.

praiskup commented:
Sorry for closing, that was inadvertently ...

trepik commented:
- changed the prefix from concit_ to ct_
- all the functions are now documented
- timeout in the ct_assert_container_creation_fails increased to 30s
- ct_wait_for_cid has an argument instead of using a global variable
- ctest_ prefix is used for single tests
- scl_usage name stayed as before
- run_doc_test renamed to ctest_doc_content
- run_all_tests renamed to run_test_list
I suggest labeling some of the functions as "use at your own risk" or "deprecated" but keep them until a new better ones are implemented

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,199 @@
+  timeout -s SIGTERM --preserve-status 30s docker run --rm "$@" $IMAGE_NAME
+  ret=$?
+  set -e
+
```
I think that if `docker run` fails with exit status `1`, this function is going to succeed.. which seems to be a bug (confirmed by @hhorak).

trepik replied:
Well I think the function should fail only if the docker run command is terminated by the timeout command... meaning that with the container_args provided the container starts properly

hhorak replied:
I'm puzzled now :) First of all, I though I already commented here, but now I don't see any comment from myself, so I might not send it.. so again..
Anyway, @praiskup, I might have been wrong when we talked about it.. IMO the function was meant to test whether the container does not start with invalid parameters. So, the behaviour should be this:
* if the container is still running after 30s --> function should fail
* if the container ends with 0 --> function should fail
* if the container ends with 1 --> function should succeed

What we might also consider (but I'm not sure whether it would bring some benefits), is running `docker inspect` and see `State.Running` and `State.ExitCode`. But maybe we're fine with the timeout above.

When I tried it myself a bit now, the only thing that did not work for me was that sometimes `SIGTERM` might not be enough.. I had to use `-s SIGKILL` to kill some daemons..

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,199 @@
+  : "Created container $(cat $cid_file)"
+}
+
+# ctest_scl_usage [name, command, expected]
```
If this is the final test-case (`ctest_` prefix), avoiding arguments might be a better idea.

praiskup replied:
That's because you can not simply list the function in $TEST_LIST, but you have to (somehow) specify the arguments, too.

trepik replied:
should I change the prefix to ct_ then? I don't get it

praiskup replied:
Well, final test functions (those which can be listed in TEST_LIST) should be named `ctest_`, IMO.

praiskup commented on test-lib.sh:
```diff
@@ -0,0 +1,199 @@
+
+# ctest_doc_content [strings]
+# --------------------
+# Looks for occurence of stirngs in the documentation files and checks
```
Similarly here, if this is "end-test" -- I'm not sure it should be parametrized;  that is -> `ctest_doc_content` function should be run against our all images, and guarantee that our images contain the same set of documentation files.

praiskup commented:
Looks much better to me, thanks!

On Thursday, August 17, 2017 4:00:44 PM CEST Tomas wrote:
> - ctest_ prefix is used for single tests

The tests should be IMO written so they don't accept arguments.  If they need
arguments, perhaps that's rather library call.

> - run_all_tests renamed to run_test_list

I would use the `ct_run_all_tests`.

> I suggest labeling some of the functions as "use at your own risk" or
> "deprecated" but keep them until a new better ones are implemented

Agreed, if you care enough please do.  E.g. `run_all_tests` was just an
work-around at the time when we were not allowed to do refactoring of testsuite.
I'll take a look at better mechanism once we have this PR merged.

praiskup commented:
Another note:  Consider that some of the methods are also useful for writing `build.sh` or `tag.sh` files, or some per-container build script in remote git repository;  IOW consider that some method is not strictly related to testing..   I'm curious whether we shouldn't rather go out with `cbl_` prefix (**general** container *build* library), while keeping the `ctest_*` prefix for the testcases.

trepik commented:
- re-implemented `ct_assert_container_creation_fails` function without using `timeout`, inspecting rather the Status and ExitCode like @hhorak suggested 
- local variable in `cleanup` function decapitalized
- `ctest_` prefixed functions renamed

praiskup commented:
[test]

omron93 commented on test-lib.sh:
```diff
@@ -0,0 +1,215 @@
+      : "Dumping logs for $container"
+      docker logs $container
+    fi
+    docker rm $container
```
@trepik @praiskup What about removing also volumes? (`docker rm -v`)

Running tests several times creates a lot of unusefull stuff. For example older versions of mongodb preallocate ~300MB with each start.

praiskup replied:
+1, but I would probably postpone the new features to new PRs.

praiskup replied:
Ah, if that's only about `-v` option, it might work well...

trepik replied:
good idea, I'll add the -v option to the `ct_cleanup` function as well

omron93 commented on test-lib.sh:
```diff
@@ -0,0 +1,215 @@
+function ct_create_container() {
+  local cid_file="$CID_FILE_DIR/$1" ; shift
+  # create container with a cidfile in a directory for cleanup
+  docker run ${CONTAINER_ARGS:-} --cidfile="$cid_file" -d "$@" $IMAGE_NAME
```
Shouldn't the `$@` be placed after `$IMAGE_NAME`? Now `$@` and `${CONTAINER_ARGS:-}` do the same and there is no way to specify command for new container.

trepik replied:
I didn't realize what the $@ part was meant to be used to. I'l fix that as you suggest. I'll also move `${CONTAINER_ARGS:-}` right infront of the `$IMAGE_NAME`

omron93 commented:
Thanks for the work done in this PR.

Rebase is needed to make tests working. On the other hand, without PR for some container which add using of these functions, testing of this PR makes no sense.

trepik commented:
I'm working on PR for mongo that would use these functions.

praiskup commented:
@trepik can you reference the PR against mongodb in commit-message, like:

    Required-by: mongodb-container#<PR_ID>

See #23 PR for more info.  Btw., seems like we should document "how to contribute".

praiskup commented:
[test]

praiskup commented:
Ah, this still needs a rebase.  But since Tomas is not going to continue on this PR (most probably(, we should submit a new PR starting from what has been done here.

praiskup commented:
Closing this, and continuing in #27.  Thanks!
